File: Data_Structures/Heaps/task_schedular.c
// Given a char array representing tasks CPU need to do, where each task is represented by a character
// and has a non-negative cooling interval n, find the least number of intervals the CPU needs to finish
// all tasks.

#include <stdio.h>
#include <stdlib.h>

int leastInterval(char* tasks, int n) {
    int charCounts[26] = {0};
    int maxCount = 0;
    int maxCountTasks = 0;

    for (int i = 0; tasks[i] != '\0'; i++) {
        charCounts[tasks[i] - 'A']++;
        if (charCounts[tasks[i] - 'A'] == maxCount)
            maxCountTasks++;
        else if (charCounts[tasks[i] - 'A'] > maxCount) {
            maxCount = charCounts[tasks[i] - 'A'];
            maxCountTasks = 1;
        }
    }

    int partitions = maxCount - 1;
    int emptySlots = partitions * (n - maxCountTasks + 1);
    int availableTasks = sizeof(tasks) / sizeof(tasks[0]) - maxCount * maxCountTasks;
    int idleSlots = (emptySlots - availableTasks) > 0 ? (emptySlots - availableTasks) : 0;

    return sizeof(tasks) / sizeof(tasks[0]) + idleSlots;
}

int main() {
    char tasks[] = {'A', 'A', 'A', 'B', 'B', 'B'};
    int n = 2;

    int minIntervals = leastInterval(tasks, n);
    printf("Least number of intervals: %d\n", minIntervals);

    return 0;
}

File: Data_Structures/Heaps/task_schedular.py
# Given a char array representing tasks CPU need to do, where each task is represented by a character
# and has a non-negative cooling interval n, find the least number of intervals the CPU needs to finish
# all tasks.

def least_interval(tasks, n):
    char_counts = [0] * 26
    max_count = 0
    max_count_tasks = 0

    for task in tasks:
        char_counts[ord(task) - ord('A')] += 1
        if char_counts[ord(task) - ord('A')] == max_count:
            max_count_tasks += 1
        elif char_counts[ord(task) - ord('A')] > max_count:
            max_count = char_counts[ord(task) - ord('A')]
            max_count_tasks = 1

    partitions = max_count - 1
    empty_slots = partitions * (n - max_count_tasks + 1)
    available_tasks = len(tasks) - max_count * max_count_tasks
    idle_slots = max(empty_slots - available_tasks, 0)

    return len(tasks) + idle_slots

tasks = ['A', 'A', 'A', 'B', 'B', 'B']
n = 2

min_intervals = least_interval(tasks, n)
print("Least number of intervals:", min_intervals)
